fix: returndata

enitrat opened the pull request:
<!--- Please provide a general summary of your changes in the title above -->

## Pull Request type

<!-- Please try to limit your pull request to one type; submit multiple
pull requests if needed. -->

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying,
or link to a relevant issue. -->

Resolves: #383

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

- Correct behavior of returndatasize/returndatacopy
- Got rid of the `child_return_data` field
- `return_data` now refers to the return_data of the last executed subcontext
- renamed machine `revert` and `stop` to `set_reverted` and `set_stopped` to avoid confusion with `revert` and `stop` opcodes
- Unexpected revert (not the REVERT Opcode) reason is not appended to the parent's return data field.


## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this does introduce a breaking change, please describe the impact and
migration path for existing applications below. -->

Eikix commented on crates/evm/src/context.cairo:
```diff
@@ -301,11 +298,6 @@ impl ExecutionContextImpl of ExecutionContextTrait {
     }
 
-    #[inline(always)]
-    fn set_return_data(ref self: ExecutionContext, value: Array<u8>) {
```
you probably still need this, 
when you call exec_return, and context is_root =>
then set_return_data as the correct slice in memory and exit the tx

Eikix commented on crates/evm/src/machine.cairo:
```diff
@@ -258,20 +259,23 @@ impl MachineCurrentContextImpl of MachineCurrentContextTrait {
+    /// the root context, does nothing.  Should be called upon returning from a
+    /// child context.
     #[inline(always)]
-    fn set_return_data(ref self: Machine, value: Array<u8>) {
```
you probably still need this, 
when you call exec_return, and context is_root =>
then set_return_data as the correct slice in memory and exit the tx

Eikix commented on crates/evm/src/instructions/system_operations.cairo:
```diff
@@ -32,9 +32,9 @@ impl SystemOperations of SystemOperationsTrait {
+        let mut return_data = Default::default();
         self.memory.load_n(size, ref return_data, offset);
-        self.set_return_data(return_data);
+        self.set_parent_return_data(return_data.span());
```
I think we need a way to differentiate if it is the root context 
In which case, write return data

Eikix commented on crates/evm/src/instructions/system_operations.cairo:
```diff
@@ -32,9 +32,9 @@ impl SystemOperations of SystemOperationsTrait {
         self.memory.load_n(size, ref return_data, offset);
-        self.set_return_data(return_data);
+        self.set_parent_return_data(return_data.span());
         self.stop();
```
why keep .stop for the machine, but not the ctx

enitrat replied:
missed it

enitrat commented:
- Updated `set_return_data` to 

```rust
    /// Sets the `return_data` field of the appropriate execution context,
    /// taking into acount EVM specs: If the current context is the root
    /// context, sets the return_data field of the root context.  If the current
    /// context is a subcontext, sets the return_data field of the parent.
    /// Should be called when returning from a context.
```

- renamed machine method to `set_stopped`

Eikix commented on crates/evm/src/instructions/system_operations.cairo:
```diff
@@ -32,10 +32,10 @@ impl SystemOperations of SystemOperationsTrait {
-        self.set_return_data(return_data);
-        self.stop();
+        self.set_return_data(return_data.span());
+        self.stopped();
```
I think you're trying to set_stopped

Maybe the exec_return test should make sure:
- the context is stopped
- the correct return data is written

enitrat replied:
thanks

Eikix commented on crates/evm/src/tests/test_utils.cairo:
```diff
@@ -244,3 +226,35 @@ fn deploy_kakarot_core() -> IExtendedKakarotCoreDispatcher {
+// Simulate return of subcontext where
+/// 1. Set `return_data` field of parent context
+/// 2. make `parent_ctx` of `current_ctx` the current ctx
+fn return_from_subcontext(ref self: Machine, return_data: Span<u8>) {
```
I wonder if we'll need this when we have the finalize_context method?
it could be used in the if (stopped) branch of the interpreter, and in the tests that need to go up one level?

this might be good though, to only target the behaviour we want,

in any case, i think its ok to have it rn, we just need to make sure it does exactly what is expected when a context is stopped

Eikix commented on crates/evm/src/tests/test_instructions/test_system_operations.cairo:
```diff
@@ -21,14 +23,15 @@ fn test_exec_return() {
 
     // Then
-    assert(1000 == load_word(32, machine.return_data()), 'Wrong return_data');
+    assert(1000 == load_word(32, parent_ctx_return_data(ref machine)), 'Wrong return_data');
```
can we call machine.parent_ctx_return_data?
